lib/tests/formulae.rb

```diff
@@ -39,6 +41,80 @@ def run!(args:)
 
       private
 
+      GRAPHQL_QUERY = <<~GRAPHQL
+        query ($node_id: ID!) {
+          node(id: $node_id) {
+            ... on Commit {
+              checkSuites(last: 100) {
+                nodes {
+                  status
+                  workflowRun {
+                    databaseId
+                    event
+                    workflow {
+                      name
+                    }
+                  }
+                  checkRuns(last: 100) {
+                    nodes {
+                      name
+                      status
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      GRAPHQL
+
+      def download_previously_built_bottles!
+        return if ENV["GITHUB_ACTIONS"].blank?
+
+        github_event_path = ENV.fetch("GITHUB_EVENT_PATH")
+        event_payload = JSON.parse(File.read(github_event_path))
+        return if event_payload.fetch("action") != "synchronize"
+
+        previous_sha = event_payload.fetch("before")
+        repo = ENV.fetch("GITHUB_REPOSITORY")
+        url = GitHub.url_to("repos", repo, "commits", previous_sha)
+        response = GitHub::API.open_rest(url)
+        node_id = response.fetch("node_id")
```

Comment:
I think I need some clarifying on the situation here.

We want to download previous bottles when we do a re-run of the job. I think we all agree that.

But this seems to be all about tracking new pushes. This sounds very unsafe as is. With a new push we may have changed the formulae, and this would just download the bottle from the previous push rather than actually rebuilding.

If we really want to have this work across new pushes rather than just re-runs, it's going to have to be a lot more intelligent by comparing the list of files changed between push attempts and invalidating all bottles from changed formulae and from any formula that has a changed dependency. This needs to track rebases too. There are severe security implications if we don't do this properly.

Reply:
Will add handling for tracking changes in formulae or dependencies.

While tracking reruns is straightforward, I'm not actually sure that GitHub exposes an API for downloading artifacts from earlier run attempts. I haven't looked _that_ hard yet, though, so it's possible I just missed it. But a cursory look suggests they only support downloading artifacts from the latest attempt. Not that this makes it impossible to do, though -- we'd just have to scrape HTML or whatever to do it, maybe.

Reply:
Ah yes, I forget it's keyed on run ID so carrying state from previous run attempts is not supported on GitHub Actions alone. Pretty sure even the internal API URLs for actions/download-artifact also do not change on new run attempts.

Reply:
For webpages: the artifacts section gets deleted from previous run attempts.

Reply:
Looking into it more, take this as an example: https://github.com/Homebrew/homebrew-core/actions/runs/5005964050. The logs for both runs are there. The artifact ID of the `bottles` artifact looks to be the same ID as the original run (701723987) given it's a lower number than the run attempt 1 logs.

I guess technically if the artifact ID can be fetched then that might be able to be used, assuming it's not in a inconsistent state.

Reply:
Think it's worth investing in trying to access artifacts from older runs? Or should we focus on validating bottles from earlier pushes? CC @MikeMcQuaid for thoughts.

Reply:
It's an interesting question as logic for both sounds like it would be almost completely separate, so do we prioritise one way or try support both? Different security risks as well.

For older run attempts, it's definitely safer security wise and probably the most useful I reckon but it's not 100% clear what's possible yet.

For earlier pushes: ideally we wouldn't incentivise having to repush when a re-run could have been appropriate. But otherwise it seems getting the artifacts is easy enough, and the challenge here is making sure validation is bulletproof as it's a vulnerability for shipping malicious bottles if not.

Reply:
> Think it's worth investing in trying to access artifacts from older runs?

If it's possible: yes and seems more useful.

> For earlier pushes: ideally we wouldn't incentivise having to repush when a re-run could have been appropriate.

Agreed.

> But otherwise it seems getting the artifacts is easy enough, and the challenge here is making sure validation is bulletproof as it's a vulnerability for shipping malicious bottles if not.

Agreed.

Simplest version to me seems to be to name the actual bottles something like `${NORMAL_BOTTLE_FILENAME}-${HASH_OF_FORMULA}.tar.gz`. Could instead use the bottle's tab because it actually has the version of all the dependencies it was built against.

Reply:
Added a check to make sure we're not installing stale bottles. It's a lot simpler than discussed above, which makes me suspicious about correctness, so let me know if it's actually mistaken.

Reply:
@carlocab I haven't given it a super thorough check but: this approach makes sense to me, better to rely on Git here.

Reply:
Sure, np. Let me sleep on it to make sure I'm happy with it.